### PR TITLE
Stream To Pipe Unit tests and Bugfix ref #1217

### PR DESCRIPTION
**Pull request by Drawaes**
Doing these tests it's highlighted a mismatch in the behavior. The CopyToAsync completes the pipe however this is not the same behavior as the stream side. I think we might what to change this behavior thoughts?

### CODE REVIEW

**Comment by Drawaes** [comment]
cc @halter73 @davidfowl

**Comment by halter73** [comment]
I would change CopyToAsync to not complete the pipe since we now have CopyToEndAsync.

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 1]
```diff
@@ -88,7 +87,7 @@ public static Task CopyToEndAsync(this IPipeReader input, Stream stream)
                 input.Complete(ex);
```
Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Shouldn't AsPipelineWriter() call this method instead of CopyToAsync() so exceptions are handled?

**Comment by davidfowl** [comment]
@Drawaes tests are failing and there are 9 commits. Can you fix the PR to:
- Be on top of master
- Fix whatever is failing

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new MemoryStream(buffer))
+            {
+                var pipe = pipeFactory.Create();
+                await stream.CopyToAsync(pipe.Writer);
+                var reader = await pipe.Reader.ReadAsync();
+
+                Assert.False(reader.IsCompleted);
+                Assert.Equal<byte>(buffer, reader.Buffer.ToArray());
```
Comment (davidfowl) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: nit: get rid of the generic arg.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Pipelines.Tests
+{
+    public class PipeToStreamFacts
+    {
+        [Fact]
+        public async Task CopyFromPipeToEnd()
```
Comment (davidfowl) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: The test name should have the expected behavior. CopyFromPipeToEndThrowsWhen???

**Comment by ahsonkhan** [comment]
> Yolo test rerun
FYI: You can re-run the tests by commenting: "@dotnet-bot test this" or specify which tests to rerun: "@dotnet-bot test Windows Debug" You shouldn't need to re-commit unnecessarily.

**Comment by Drawaes** [comment]
Good to know

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 4]
```diff
@@ -81,41 +81,54 @@ public static Task CopyToEndAsync(this IPipeReader input, Stream stream)
         {
             try
             {
-                await input.CopyToAsync(stream, bufferSize, cancellationToken);
+                // TODO: Use bufferSize argument
+                while (!cancellationToken.IsCancellationRequested)
```
Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: If this method is going to take a CancellationToken, it makes sense to register a cancellation callback that calls CancelPendingRead() in case the IPipeReader is idle.

Comment (davidfowl) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Lets do that in another PR.

Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Also, should input.Complete(TCE) be called if a cancellation is requested during the copy?

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 5]
```diff
@@ -81,41 +81,54 @@ public static Task CopyToEndAsync(this IPipeReader input, Stream stream)
...
+
+                        await inputBuffer.CopyToAsync(stream);
+                    }
+                    finally
+                    {
+                        input.Advance(inputBuffer.End);
+                    }
+                }
             }
             catch (Exception ex)
```
Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: I'm not sure it makes as much sense to swallow exceptions when copying to a Stream since the implementer of the stream can't observe the exception when IPipeReader.ReadAsync() throws.

There's also a decent chance the producer won't observe exceptions from Stream.WriteAsync() either if it throws after all the produced data has already been "flushed".

What do you think @davidfowl?

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs (hunk 3 above)
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: Agreed.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new FakeExceptionStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                pipe.Writer.Complete();
+                await pipe.Reader.CopyToEndAsync(stream);
+                await Assert.ThrowsAsync<InvalidOperationException>(async () =>
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: CopyFromPipeToEndObserveStreamThrow doesn't seem like the appropriate name for what this is testing.

If this was really observing the stream throwing, wouldn't this be an IOException? It seems more like CopyToEndAsync called `IPipeReader.Complete(IOException)`, the the producer was no longer around to observe it. This is a good example of the issue I described in my comment in PipelineConnectionExtensions.cs.

What's really happening here is that `IPipeReader.ReadAsync()` can't be called after  `IPipeReader.Complete()` is with or without an exception.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+        {
+            var buffer = new byte[1000];
+            (new Random()).NextBytes(buffer);
+            using (var pipeFactory = new PipeFactory())
+            using (var stream = new MemoryStream())
+            {
+                var pipe = pipeFactory.Create();
+                await pipe.Writer.WriteAsync(buffer);
+                await pipe.Reader.CopyToAsync(stream);
+                var reader = pipe.Reader.ReadAsync();
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: I can see the argument for not calling `IPipeWriter.Complete()` when a stream is fully copied to a pipe so even more data can be copied. I don't see why you would ever not call `IPipeReader.Complete()` when the data from the pipe has been fully consumed.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+                Assert.False(reader.IsCompleted);
+                Assert.Equal<byte>(buffer, stream.ToArray());
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromStreamToEnd()
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: Like the first test, this could also use a better name. Maybe CopyFromStreamToEndCompletesPipeWriter.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+                var reader = await pipe.Reader.ReadAsync();
+
+                Assert.True(reader.IsCompleted);
+                Assert.Equal(buffer, reader.Buffer.ToArray());
+                pipe.Reader.Advance(reader.Buffer.End);
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromStreamToEndObserveStreamThrow()
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: Maybe CopyFromStreamToEndObserveStreamThrowsFromPipeReadAsync.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+                await Assert.ThrowsAsync<IOException>(async () =>
+                {
+                    var reader = await pipe.Reader.ReadAsync();
+                });
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromStreamNoEndObserveStreamThrow()
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: Maybe CopyFromStreamNoEndObserveStreamThrowsFromCopyToAsync.

File: tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                var pipe = pipeFactory.Create();
+                await Assert.ThrowsAsync<IOException>(async () =>
+                {
+                    await stream.CopyToAsync(pipe.Writer);
+                });
+            }
+        }
+
+        [Fact]
+        public async Task CopyFromStreamNoEnd()
```
Comment (halter73) [inline tests/System.IO.Pipelines.Tests/PipeToStreamFacts.cs]: Maybe CopyFromStreamNoEndDoesNotCompletePipeWriter

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 12]
```diff
@@ -81,41 +81,54 @@ public static Task CopyToEndAsync(this IPipeReader input, Stream stream)
...
+                // TODO: Use bufferSize argument
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    var result = await input.ReadAsync();
+                    var inputBuffer = result.Buffer;
+                    try
+                    {
+                        if (inputBuffer.IsEmpty && result.IsCompleted)
+                        {
+                            input.Complete();
```
Comment (davidfowl) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: I don't like this control flow. Do it at the end of the method in a finally or something.

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs (hunk 4 above)
Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Registering to the token can be done in another PR, but it still shouldn't be possible to exit this method without Completing the IPipeReader. The easiest solution for now is to remove the CancellationToken overload until it's properly supported.

As I mentioned in my comment in the "CopyFromPipeNoEnd()" test, it doesn't make sense to have separate CopyToAsync and CopyToEndAsync methods in this direction since you always finish reading the IPipeReader in either method (unless Stream.WriteAsync() throws I guess). Might as well call IPipeReader .Complete().

Even when Stream.WriteAsync() throws, a lot of the time the producer will no longer be producing data making it unable to observe the exception as is the case in the "CopyFromPipeToEndObserveStreamThrow" test.

File: src/System.IO.Pipelines/PipelineConnectionExtensions.cs [hunk 13]
```diff
@@ -81,41 +81,54 @@ public static Task CopyToEndAsync(this IPipeReader input, Stream stream)
...
         public static Task CopyToAsync(this IPipeReader input, Stream stream)
         {
-            return input.CopyToAsync(stream, 4096, CancellationToken.None);
+            return input.CopyToAsync(stream, 4096);
         }
 
-        public static async Task CopyToAsync(this IPipeReader input, Stream stream, int bufferSize, CancellationToken cancellationToken)
+        public static async Task CopyToAsync(this IPipeReader input, Stream stream, int bufferSize)
         {
-            // TODO: Use bufferSize argument
```
Comment (halter73) [inline src/System.IO.Pipelines/PipelineConnectionExtensions.cs]: Why remove this comment? bufferSize is still unused in this method right?  Can we just remove the bufferSize argument?

**Comment by Drawaes** [comment]
These methods are moving so let's close this change out
